src/main/java/org/broadinstitute/hellbender/utils/bwa/BwaMemIndex.java

```diff
@@ -0,0 +1,349 @@
+package org.broadinstitute.hellbender.utils.bwa;
+
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.CigarElement;
+import htsjdk.samtools.CigarOperator;
+import htsjdk.samtools.SAMFlag;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.NativeUtils;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.List;
+
+public final class BwaMemIndex implements AutoCloseable {
+    private long indexAddress;
+    private final List<String> refContigNames;
+    private static volatile BwaMemIndex globalInstance;
+
+    public static void createIndexImage( final String refName, final String imgName ) {
+        assertNonEmptyReadable(refName+".amb");
+        assertNonEmptyReadable(refName+".ann");
+        assertNonEmptyReadable(refName+".bwt");
+        assertNonEmptyReadable(refName+".pac");
+        assertNonEmptyReadable(refName+".sa");
+        createIndexImageFile(refName, imgName);
+    }
+
+    public static BwaMemIndex getInstance( final String indexImageFile ) {
+        BwaMemIndex result = globalInstance;
+        if ( result == null ) {
+            synchronized(BwaMemIndex.class) {
+                result = globalInstance;
+                if ( result == null ) globalInstance = result = new BwaMemIndex(indexImageFile);
+            }
+        }
+        return result;
+    }
+
+    public static void closeInstance() {
+        BwaMemIndex result = globalInstance;
+        if ( result != null ) {
+            synchronized(BwaMemIndex.class) {
+                result = globalInstance;
+                if ( result != null ) {
+                    globalInstance = null;
+                    result.close();
+                }
+            }
+        }
+    }
+
+    public static void closeAllDistributedInstances( final JavaSparkContext ctx ) {
+        int nJobs = ctx.defaultParallelism();
```

Comment:
Does this actually guarantee that it will run on every executor?  It seems plausible that it may end up running several partitions on the same executor, especially if it's a quick operation.   Do you know?